<h1> <center> Programmation Python - Les bases </center></h1>

<p style="text-align:justify;">On présente ici les bases du langage <strong>Python</strong> : calculs, variables, affichage, saisie, type de données.
Les fonctionnalités un peu plus avancées telles que les instructions conditionnelles, les boucles, les fonctions ou les listes seront étudiées dans de prochaines séances.</p>

<table> <tr> <td width="250"> <img src="https://capytale2.ac-paris.fr/web/sites/default/files/2022/09-01/19-07-33/800px-Guido-portrait-2014-drc.jpg"/> </td>
<td> <div style="text-align:left;"> <big> Python est un langage de programmation, développé par Guido Van Rossum en 1989. Il est libre et gratuit, fonctionne sur tout type de plateformes informatiques et de systèmes d’exploitation (Windows, Mac OS, Unix, Linux, Android…). C’est un langage de programmation interprété (pas besoin de compilateur), et multi-paradigmes dans le sens où il permet différents types de programmation (impérative, fonctionnelle, orientée objet). Il offre une grande variété d’outils pour des usages simples ou complexes, il est extensible, sa syntaxe est claire et cohérente ; ce qui en fait un langage apprécié pour l’initiation aux concepts de la programmation et très largement répandu.
</big></div></td></tr></table>

<div style="padding:10px ;border:solid 2px lightcoral; border-radius: 10px; background-color:bisque;">
    Dans la suite vous devez exécuter les cellules de code <b>Entrée[ ]</b> en cliquant sur <button class='fa fa-step-forward'></button> &laquo; Exécuter &raquo;

<div style="padding:10px ;border:solid 2px lightblue; border-radius: 10px; background-color:aliceblue;">
N'oubliez pas de sauvegarder régulièrement votre travail : cliquer sur <button class='fa fa-save icon-save btn btn-xs btn-default'></button> !</div>

<h2> Partie 1. Nombres et calculs </h2>

Python permet d'effectuer tout type de calcul, comme une calculatrice ordinaire.

In [1]:
5 + 8

13

Les règles de priorité des calculs sont les mêmes qu'en mathématiques. On utilise des parenthèses lorsqu'on a besoin de changer l'ordre de priorité. Le symbole de la multiplcation est * (afin qu'il n'y ait pas de confusion avec la lettre x).

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 1. </strong> Dans une école, il y a trois classes : <ul>
    <li>Classe A : 32 élèves</li>
    <li>Classe B : 27 élèves</li>
    <li>Classe C : 34 élèves</li>
</ul>  <br> 
Chaque élève de l'école doit recevoir 192 photocopies.<br>
Combien faut-il prévoir de photocopies en tout ?
</div>

In [9]:
Classe_A = 32 
Classe_B = 27
Classe_C = 34

photocopies_par_eleve = 192

total_photocopies = (Classe_A + Classe_B + Classe_C) * photocopies_par_eleve

print(total_photocopies)

17856


<p> &#x270E; Phrase réponse : il faut prevoir 17856 photocopies. </p>

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 2. </strong> Calculer le produit de deux nombres entiers à 10 chiffres.
</div>

In [10]:
nombre1 = 12
nombre2 = 22

produit = nombre1 * nombre2
print(produit)

264


Pour calculer une <strong>puissance</strong>, on utilise la notation ** . Par exemple pour calculer 7 au carré, on écrira 7**2.

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 3. </strong> Calculer $6^3$ de deux manières différentes : en utilisant une puissance et sans utiliser de puissance. 
</div>

In [11]:
resultat_puissance = 6**3
print(resultat_puissance)

216


In [12]:
resultat_sans_puissance = 6 * 6 * 6
print(resultat_sans_puissance)

216


Il existe deux types de division en Python : la division entière avec l'opérateur // qui permet d'obtenir le dividende et la division flottante avec l'opérateur / .<br>
L'opérateur % permet d'obtenir le reste dans une division entière.

In [13]:
13 // 5

2

In [14]:
13 / 5

2.6

In [15]:
13 % 5

3

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 4. </strong> En utilisant les opérateurs précédents, écrire 437 secondes au format minutes:secondes.
</div>

Nombre de minutes : 

In [23]:
437//60

7

Nombre de secondes :

In [24]:
437%60

17

**À retenir. Les opérateurs de bases**

| Opération       | Caractère   |
| :-------------- | :---------: |
| Addition        |           + |
| Soustraction    |           - |
| Multiplication  |          \* |
| Division        |           / |
| Division entière|          // |
| Reste           |           % |
| Puissance       |        \*\* |

On rappelle que l'écriture scientifique d'un nombre décimal est de la forme $d \times 10^{k}$ où $d$ est un nombre décimal positif ou négatif comportant un seul chiffre avant la virgule et $k$ est un entier positif ou négatif.

Par exemple, l'écriture scientifique de $-0,0042$ est $-4,2 \times 10^{-3}$.

En Python, on peut utiliser l'écriture scientifique des nombres décimaux, par exemple $-4,2 \times 10^{-3}$ s'écrit

In [25]:
-4.2e-3

-0.0042

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 5. </strong> Donner l'écriture scientifique en Python des nombres décimaux suivants et vérifier votre résultat.
</div>

Écriture scientifique de 227,52 :

In [26]:
2.2752e2

227.52

Écriture scientifique de 0,000127 :

In [27]:
1.27e-4

0.000127

Le nombre $\pi$ est défini dans la bibliothèque ou module math, il faut l'importer avant de pouvoir l'utiliser.

In [28]:
from math import pi
pi

3.141592653589793

De même, si on veut utiliser les racines carrées d'un nombre, il faut importer la fonction racine carrée **sqrt** du module math.

In [29]:
from math import sqrt
sqrt(25)

5.0

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 6.</strong> Soit ABC un triangle rectangle en A avec AB = 5 et BC = 7. 
    <ol type="a">
        <li> Donner une valeur approchée à $10^{-2}$ près de AC.</li>
        <li> Donner une valeur approchée à $10^{-2}$ près de la circonférence du cercle de diamètre BC.</li>
    </ol>
</div>

In [37]:
AB = 5
BC = 7

AC = sqrt(AB**2 + BC**2)

print("La longueur de AC est :", round(AC, 2))


La longueur de AC est : 8.6


<p> &#x270E; Phrase réponse : la longueur BC vaut environ 8.6 </p>

In [38]:

from math import pi

diametre_BC = 7

circonference = pi * diametre_BC

print(round(circonference, 2))



21.99


<p> &#x270E; Phrase réponse : la circonférence du cercle de diamètre BC vaut environ 21.99 </p>

Il existe deux types de nombres en Python : les <strong>entiers</strong> (<code>int</code>, abréviation de "integer") et les <strong>flottants</strong> (<code>float</code>, abréviation de "floating number" qui signie "nombre à virgule flottante"). 

La fonction <code>type()</code> indique le type du nombre qu'on lui donne en paramètre.

In [39]:
type(4)

<class 'int'>

In [40]:
type(4.0)

<class 'float'>

<h2> Partie 2. Variables et affectation</h2>

<div style="display:inline-block">
    Voici un algorithme écrit en langage naturel. Dans ce langage, on utilise le symbole $\gets$ pour l'affectation d'une variable. En Python, on utilisera le signe =.
$$\textrm{Algorithme}\\
\begin{array}{lcl}
a & \gets & 2 \\
b & \gets & 3 \\
a & \gets & b+1 \\
b & \gets & 5 \\
\end{array}$$
    On le programme en langage Python.
</div>

In [42]:
a = 2
b = 3
a = b + 1
b = 5

Quelles sont les valeurs finales de $a$ et $b$ ? Vérifier à l'aide de la cellule ci-dessous. 

In [49]:
print (a, b)

4 5


Afficher une variable qui n'a pas encore été définie crée une erreur.

In [47]:
print (d)

Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'd' is not defined


Error: 

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 7. </strong> Prédire ce que fait l'algorithme ci-dessous, puis le programmer en langage Python pour vérifier les prédictions.
$$\textrm{Algorithme}\\
\begin{array}{lcl}
a & \gets & 2 \\
b & \gets & 3 \\
a & \gets & b \\
b & \gets & a \\
\end{array}$$
</div>

<p> &#x270E; Prédiction :  a aura la valeur 23 et b aura également la valeur 23 (puisque b est assigné à a).</p>

In [53]:
#Code
a = 23  
b = a   



In [54]:

# Afficher les valeurs finales de a et b

print("Valeur de a:", a)
print("Valeur de b:", b)

Valeur de a: 23
Valeur de b: 23


<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 8. </strong> Compléter le programme ci-dessous pour qu'il échange le contenu des variables a et b (3 lignes à écrire).
</div>

In [55]:
a = 7
b = 4

temp = a  
a = b    
b = temp  



Valeur de a: 4
Valeur de b: 7


Vérification

In [57]:

print("Valeur de a:", a)
print("Valeur de b:", b)

Valeur de a: 4
Valeur de b: 7


<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 9. </strong> Que fait le programme ci-dessous ? <br>
    <code> a = 7 </code> <br>
    <code> b = 4 </code> <br>
    <code> a = a + b </code> <br>
    <code> b = a - b </code> <br>
    <code> a = a - b </code>
</div>

<p> &#x270E; Phrase réponse : Le programme échange les valeurs de a et b, des sorte que a prend la valeur de 4 et b prend la valuer de 7 </p>

In [62]:
a = 7
b = 4
a = a + b
b = a - b
a = a - b

Vérification

In [61]:
a = 7
b = 4
a = a + b
b = a - b
a = a - b

print(a, b) 

4 7


**Preuve.** Compléter le tableau de suivi des variables ci-dessous (remplacer les ? par les valeurs obtenues).

| Instructions | valeur de a | valeur de b |
| :----------- | :---------: | :---------: |
| a = 7        | 7           |             |
| b = 4        | 7           | 4           |
| a = a + b    | ?           | ?           |
| b = a - b    | ?           | ?           |
| a = a - b    | ?           | ?           |

<h2> Partie 3. Les chaînes de caractères </h2>

<p style="text-align:justify;">La fonction <code>print()</code> permet d'afficher une chaîne de caractères ("string" en anglais).
La chaîne de caractère que l'on souhaite afficher doit être écrite entre guillemets (simples ' ou doubles "). Sans guillemets, c'est le contenu de la variable dont le nom est transmis en paramètre qui sera affiché.</p>

In [63]:
bonjour = 8
print("bonjour")
print(bonjour)

bonjour
8


<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 10. </strong> Écrire une ligne de programme qui affiche le message &laquo; J'ai tout compris ! &raquo;. 
</div>

In [64]:
print("j'ai tout compris ! ")

j'ai tout compris ! 


Une variable peut contenir une chaîne de caractères.

In [65]:
message = "Hello"
print(message)
type(message)

Hello


<class 'str'>

On peut concaténer (fusionner) plusieurs chaînes de caractère au moyen de l'opérateur <code>+</code> :

In [66]:
sujet = "Alice "
verbe = "aime "
cod = "les glaces"
point = "."
phrase = sujet + verbe + cod + point
print(phrase)

Alice aime les glaces.


On ne peut pas concaténer une chaîne de caractères avec une valeur numérique directement. Il faut au préalable transformer la valeur numérique en chaîne de caractères au moyen de la fonction <code>str()</code> :

In [67]:
age = 25
début = "J'ai "
fin = " ans."
phrase = début + str(age) + fin
print(phrase)

J'ai 25 ans.


Les chaines de caractères formatées (aussi appelées f-strings) permettent d’inclure la valeur d’expressions Python dans des chaines de caractères en les préfixant avec f et en écrivant les expressions entre accolades.

In [68]:
age = 21
print(f"J'ai {age} ans.")

J'ai 21 ans.


<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 11. </strong> Le nombre $\dfrac{1+\sqrt{5}}{2}$ s'appelle le nombre d'or. <br>
    Calculer une valeur approchée de ce nombre et afficher une phrase réponse du type &laquo; Le nombre d'or vaut environ ... &raquo; en utilisant une chaîne de caractère formatée.
</div>

In [70]:
import math

sqrt_5 = math.sqrt(5)

golden_ratio = 1 + sqrt_5

print("Le nombre d'or vaut environ", golden_ratio)

Le nombre d'or vaut environ 3.23606797749979


<h2> Partie 4. Saisie d'une entrée par l'utilisateur </h2>

On peut demander à l'utilisateur de saisir une valeur avec la fonction <code>input()</code>. On peut ajouter en paramètre facultatif un message qui donne des précisions sur la valeur attendue.

In [ ]:
a = input()
b = input("Entrez votre âge :")

La valeur renvoyée par la fonction <code>input()</code> est du type chaîne de caractères (string). Si on souhaite utiliser cette valeur pour effectuer des calculs, on doit d'abord convertir cette chaîne en entier avec la fonction <code>int()</code> ou en flottant avec la fonction <code>float()</code>.

In [ ]:
saisie = input("Veuillez entre votre âge : ")
age = int(saisie)
age = age + 5
print(f"Dans 5 ans, vous aurez {age} ans.")

Si on veut saisir des nombres définis par une expression algébrique comme $21+5$ ou $\sqrt{2}$ et les utiliser pour un calcul, les fonctions respectives <code>int()</code> ou <code>float()</code> ne fonctionnent pas.  

In [ ]:
saisie = input("Entrer une somme de deux entiers : ")
nombre = int(saisie)
print(f"La somme vaut {nombre}.")

In [ ]:
saisie = input("Entrer la racine carrée d'un entier : ")
nombre = float(saisie)
print(f"Une valeur approchée de cette racine est {nombre}.")

Pour saisir une expression algébrique et qu'elle soit interprétée comme un nombre, il faut utiliser la fonction <code>eval()</code>.

In [ ]:
saisie = input("Entrer une somme de deux entiers : ")
nombre = eval(saisie)
print(f"La somme vaut {nombre}.")

In [ ]:
saisie = input("Entrer la racine carrée d'un entier : ")
nombre = eval(saisie)
print(f"Une valeur approchée de cette racine est {nombre}.")

<div class="alert alert-warning" style="border-left: 3px solid">
    <strong>Exercice 12. </strong> Écrire un programme qui demande à l'utilisateur d'entrer un nombre $x$ et qui calcule le nombre $3x^2 -2x+5$.<br>
<em>Le programme doit fonctionner pour n'importe quel nombre et devra afficher une phrase de réponse complète se terminant par un point</em>.
</div>

In [1]:
x = float(input("Entrez un nombre x : "))

resultat = 3 * (x ** 2) - 2 * x + 5

print("Le résultat de l'expression 3x² - 2x + 5 pour x =", x, "est", resultat, ".")

Entrez un nombre x : 5
Le résultat de l'expression 3x² - 2x + 5 pour x = 5.0 est 70.0 .


1. 
- Quel est le reste de la division euclidienne de 16 par 5 ? 
- Quel est le reste de la division euclidienne de 15 par 5 ? 
- En déduire un critère pour déterminer si un nombre _m_ est multiple d’un nombre _d_. 
2. Écrire un algorithme disant si un nombre _m_ est multiple d’un nombre _d_.
3. Coder cet algorithme en Python. _Indication :_ pour obtenir le reste de la division euclidienne de _y_ par _x_ en Python, on fait `y%x` ; par exemple `16%5` donne 1.

**Critère**

**Un algorithme:**

In [ ]:
m = int(input("Entrez un nombre m : "))  
d = int(input("Entrez un diviseur d : ")) 

if m % d == 0:
    print(m, "est un multiple de", d)
else:
    print(m, "n'est pas un multiple de", d)

<div style="padding:10px ;border:solid 2px lightcoral; border-radius: 10px; background-color:bisque;">
    N'oublier pas de rendre votre travail, cliquer en haut en droite.
</div>

<i> Cette activité a été créée en adaptant un TP de Thomas Foirien et mis à disposition sur la plateforme Capytale sous licence Creative Commons BY-NC-SA. </i>